### PR TITLE
Qubit rotation with JAX

### PR DESCRIPTION
**Pull request by araza6**
Further to https://github.com/qgrad/qgrad/issues/1, I have added the qubit rotation tutorial using JAX's `grad` on the cost function (fidelity in this case) in the gradient descent optimization routine. This hopefully gives us a better idea of how to proceed with adding autodiff capabilities in QuTiP. There are a few hiccups that we will need to tackle. I shall create separate issues for them and link them here.

### CODE REVIEW

**Comment by quantshah** [comment]
Hi, it is quite difficult to review code as .ipynb notebooks so I would say that you should make a folder called examples and put the example you write here as a .py file. Then we can easily comment and change things. Please do so and then close this PR by adding a comment to the new PR with something like:

"closing this PR via #(the new PR number)"

**Comment by nathanshammah** [comment]
I agree on the separate folder. This Review Notebook [app](https://github.com/marketplace/review-notebook-app) makes it somewhat easier to review and visualize (since many times Github jupyter notebooks visualization fails) notebooks. Maybe we could use it.

**Comment by nathanshammah** [comment]
By the way, @araza6, I like the way you listed different features in separate commits. That makes reviewing easier in general: let's stick to that.

**Comment by nathanshammah** [comment]
Actually, on "diffing" and reviewing notebooks, there is this tool within the Jupyter organization that is actively developed, https://github.com/jupyter/nbdime.

**Comment by araza6** [comment]
Thanks for the helpful comments!

- I agree with your opinion on .ipynb files in the example folder. There is a trade-off between user-experience and ease of code review. Can we do something similar to QuTiP? As in have a `qgrad-notebooks` repository under qgrad organization?

-  The motivation behind importing numpy twice (and differently each time) is because JAX's numpy, although generally faster and equipped with just-in-time compilation among other things, does not work well with QuTiP's `Qobj`; for that, we need the actual `numpy`. It should be noted, however, that JAX's numpy wrapper does not come with everything traditional numpy comes with. More [here](https://jax.readthedocs.io/en/latest/jax.numpy.html). 

As for the way libraries are imported, the import convention(s) is/are largely the same. JAX's [tutorials] (https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html) use `import numpy as onp`, whereas we, as you point out, use `import numpy as rnp`. We can probably stick to JAX's convention of onp.

- As for prefixes, you mean to `import jax` and use `jax.grad` in the notebook every time we use `grad`?

**Comment by quantshah** [comment]
Hi, I agree with the separate notebooks repo. Notebooks are nice and there are many options to allow for beautiful documentation which we should definitely provide (including easy download buttons as the user browses through notebooks). 

Since we are still tinkering around with code now, I would say we should write most of the examples as .py files and then use the nbconvert option to get to notebooks later. We could even just use sphinx gallery. This will avoid unnecessary pain later and make sure we can test the examples also.

Please note that even in .py files you can add markdown and images etc. (like this [example](https://github.com/XanaduAI/qml/blob/master/demonstrations/tutorial_data_reuploading_classifier.py) which gets rendered as https://pennylane.ai/qml/demos/tutorial_data_reuploading_classifier.html)

I have a new example using purely Jax to implement this notebook which I will soon upload as an example here.

**Comment by quantshah** [comment]
Also, @nathanshammah it is a bit confusing but @araza6 is right about the imports here. Since we need to work with two different Numpys. The easy solution would have been to somehow make a qutip fork which replaces all the `import numpy as np` with `import jax.numpy as np` but it will never work.

**Comment by nathanshammah** [comment]
There are still outstanding changes that have not been addressed, from my comments above:
- Add a title to the notebook in Markdown.

- please introduce clearly the scope of the example at the beginning of the notebook

- please comment the result on the Bloch sphere. Just like in paper, figures better be commented with some "captions". Also, comment what the different colored arrows represent.

- That's fine with the library imports for the moment, as I need to understand better how this is done. 
- PR #8 is taking care of the gitignore and should be merged first, once ready. Then we can sync this one with the origin master.

**Comment by araza6** [comment]
@nathanshammah I am including all your suggested changes in a .py file and opening a new PR as @quantshah had suggested.

**Comment by araza6** [comment]
Closing this on PR https://github.com/qgrad/qgrad/pull/10
